langchain/src/vectorstores/googlevertexai.ts

```diff
@@ -0,0 +1,682 @@
+    super(fields, caller);
+
+    this.index = fields.index;
+  }
+
+  async buildUrl(): Promise<string> {
+    const projectId = await this.auth.getProjectId();
+    const url = `https://${this.endpoint}/${this.apiVersion}/projects/${projectId}/locations/${this.location}/indexes/${this.index}:removeDatapoints`;
+    return url;
+  }
+
+  buildMethod(): string {
+    return "POST";
+  }
+
+  async request(
+    datapointIds: string[],
+    options: AsyncCallerCallOptions
+  ): Promise<RemoveDatapointResponse> {
+    const data: RemoveDatapointRequest = {
+      datapointIds,
+    };
+    return this._request(data, options);
+  }
+}
+
+interface UpsertDatapointParams extends GoogleVertexAIConnectionParams {
+  index: string;
+}
+
+export interface Restriction {
+  namespace: string;
+  allowList?: string[];
+  denyList?: string[];
+}
+
+interface CrowdingTag {
+  crowdingAttribute: string;
+}
+
+interface IndexDatapoint {
+  datapointId: string;
+  featureVector: number[];
+  restricts?: Restriction[];
+  crowdingTag?: CrowdingTag;
+}
+
+interface UpsertDatapointRequest {
+  datapoints: IndexDatapoint[];
+}
+
+interface UpsertDatapointResponse extends GoogleVertexAIResponse {
+  // Should be empty
+}
+
+class UpsertDatapointConnection extends GoogleVertexAIConnection<
+  AsyncCallerCallOptions,
+  UpsertDatapointResponse
+> {
+  index: string;
+
+  constructor(fields: UpsertDatapointParams, caller: AsyncCaller) {
+    super(fields, caller);
+
+    this.index = fields.index;
+  }
+
+  async buildUrl(): Promise<string> {
+    const projectId = await this.auth.getProjectId();
+    const url = `https://${this.endpoint}/${this.apiVersion}/projects/${projectId}/locations/${this.location}/indexes/${this.index}:upsertDatapoints`;
+    return url;
+  }
+
+  buildMethod(): string {
+    return "POST";
+  }
+
+  async request(
+    datapoints: IndexDatapoint[],
+    options: AsyncCallerCallOptions
+  ): Promise<UpsertDatapointResponse> {
+    const data: UpsertDatapointRequest = {
+      datapoints,
+    };
+    return this._request(data, options);
+  }
+}
+
+interface FindNeighborsConnectionParams extends GoogleVertexAIConnectionParams {
+  indexEndpoint: string;
+
+  deployedIndexId: string;
+}
+
+interface FindNeighborsRequestQuery {
+  datapoint: {
+    datapointId: string;
+    featureVector: number[];
+    restricts?: Restriction[];
+  };
+  neighborCount: number;
+}
+
+interface FindNeighborsRequest {
+  deployedIndexId: string;
+  queries: FindNeighborsRequestQuery[];
+}
+
+interface FindNeighborsResponseNeighbor {
+  datapoint: {
+    datapointId: string;
+    crowdingTag: {
+      crowdingTagAttribute: string;
+    };
+  };
+  distance: number;
+}
+
+interface FindNeighborsResponseNearestNeighbor {
+  id: string;
+  neighbors: FindNeighborsResponseNeighbor[];
+}
+
+interface FindNeighborsResponse extends GoogleVertexAIResponse {
+  data: {
+    nearestNeighbors: FindNeighborsResponseNearestNeighbor[];
+  };
+}
+
+class FindNeighborsConnection
```

Comment:
I think this and things above might be better in a utils class? Either the same one as the other connection logic or a new one for matching engine?

Reply:
I'm not sure putting them in a separate utils class does any benefit. All of them are specific to Matching Engine, and  are mostly done this way to address type checking in TypeScript. And each one is specific to the operation.

Reply:
Just to reduce clutter - don't feel too strongly.